libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,551 @@
+{
+	// TODO: it would be nicer to replace this by a constructive algorithm.
+	// Currently it uses a reduced version of the Heap Algorithm to partly brute-force, which seems
+	// to work decently well.
+
+	Stack commonPrefix;
+	for (auto&& [slot1, slot2]: ranges::zip_view(_stack1, _stack2))
+	{
+		if (!(slot1 == slot2))
+			break;
+		commonPrefix.emplace_back(slot1);
+	}
+
+	Stack stack1Tail = _stack1 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+	Stack stack2Tail = _stack2 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+
+	if (stack1Tail.empty())
+		return commonPrefix + compressStack(stack2Tail);
+	if (stack2Tail.empty())
+		return commonPrefix + compressStack(stack1Tail);
+
+	Stack candidate;
+	for (auto slot: stack1Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	for (auto slot: stack2Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	cxx20::erase_if(candidate, [](StackSlot const& slot) {
+		return holds_alternative<LiteralSlot>(slot) || holds_alternative<FunctionCallReturnLabelSlot>(slot);
+	});
+
+	std::map<size_t, Stack> sortedCandidates;
+
+	auto evaluate = [&](Stack const& _candidate) -> size_t {
+		size_t numOps = 0;
+		Stack testStack = _candidate;
+		auto swap = [&](unsigned _swapDepth) { ++numOps; if (_swapDepth > 16) numOps += 1000; };
+		auto dupOrPush = [&](StackSlot const& _slot)
+		{
+			if (canBeFreelyGenerated(_slot))
+				return;
+			auto depth = util::findOffset(ranges::concat_view(commonPrefix, testStack) | ranges::views::reverse, _slot);
+			if (depth && *depth >= 16)
+				numOps += 1000;
+		};
+		createStackLayout(testStack, stack1Tail, swap, dupOrPush, [&](){} );
+		testStack = _candidate;
+		createStackLayout(testStack, stack2Tail, swap, dupOrPush, [&](){});
+		return numOps;
+	};
+
+	// See https://en.wikipedia.org/wiki/Heap's_algorithm
+	size_t n = candidate.size();
+	sortedCandidates.insert(std::make_pair(evaluate(candidate), candidate));
+	std::vector<size_t> c(n, 0);
+	size_t i = 1;
+	while (i < n)
+	{
+		if (c[i] < i)
+		{
+			if (i & 1)
+				std::swap(candidate.front(), candidate[i]);
+			else
+				std::swap(candidate[c[i]], candidate[i]);
+			sortedCandidates.insert(std::make_pair(evaluate(candidate), candidate));
+			++c[i];
+			// Note that for a proper implementation of the Heap algorithm this would need to revert back to ``i = 1.``
+			// However, the incorrect implementation produces decent result and the proper version would have n!
+			// complexity and is thereby not feasible.
+			++i;
+		}
+		else
+		{
+			c[i] = 0;
+			++i;
+		}
+	}
+
+	return commonPrefix + sortedCandidates.begin()->second;
+}
+
+Stack StackLayoutGenerator::compressStack(Stack _stack)
+{
+	optional<size_t> firstDupOffset;
+	do
+	{
+		if (firstDupOffset)
+		{
+			if (_stack.size() - *firstDupOffset - 1 > 1)
+				std::swap(_stack.at(*firstDupOffset + 1), _stack.back());
```

Comment:
Why is this needed?

Reply:
Hm... the idea was that this does something similar to ``createIdealLayout`` in small, i.e. it creates a compressed pre-layout, s.t. moving from that back to the original layout is cheap... but I'd need to think about if and why this actually helps for that - let me see what happens if I just remove it :-).

Reply:
Ah well, as it is with most of these things: partly better partly worse... but with a tendency of getting better I think. I'll think this through once again, when I did just now, I didn't see why this should help anymore, maybe I'll just remove it.

Reply:
Ok, I had another look at it - it does have weird effects on gas, both good and bad, but I think generally it should be removed and was mainly a workaround for older versions of the stack shuffling (when that didn't have ``bringUpTargetSlot`` yet). So I removed it.